Makefile

```diff
@@ -347,87 +282,155 @@ docker-push-dns-server:
 docker-push-chaos-kernel:
 	docker push "${DOCKER_REGISTRY_PREFIX}pingcap/chaos-kernel:${IMAGE_TAG}"
 
-$(GOBIN)/controller-gen:
-	$(GO) get sigs.k8s.io/controller-tools/cmd/controller-gen@v0.4.1
-$(GOBIN)/revive:
-	$(GO) get github.com/mgechev/revive@v1.0.2-0.20200225072153-6219ca02fffb
-$(GOBIN)/failpoint-ctl:
-	$(GO) get github.com/pingcap/failpoint/failpoint-ctl@v0.0.0-20200210140405-f8f9fb234798
-$(GOBIN)/goimports:
-	$(GO) get golang.org/x/tools/cmd/goimports@v0.1.4
-$(GOBIN)/gosec:
-	$(GO) get github.com/securego/gosec/cmd/gosec@v0.0.0-20200401082031-e946c8c39989
-$(GOBIN)/gqlgen:
-	$(GO) get github.com/99designs/gqlgen@v0.13.0
-
-lint: $(GOBIN)/revive
-	@echo "linting"
-	$< -formatter friendly -config revive.toml $$($(PACKAGE_LIST))
-
 bin/chaos-builder:
 	$(CGOENV) go build -ldflags '$(LDFLAGS)' -o bin/chaos-builder ./cmd/chaos-builder/...
 
 chaos-build: bin/chaos-builder
 	bin/chaos-builder
 
-generate-ctrl: $(GOBIN)/gqlgen
-	go generate ./pkg/ctrlserver/graph; make fmt;
+define RUN_IN_DEV_ENV_TEMPLATE
+ifeq ($$(IN_DOCKER),1)
+$(1):$(2)
+$($(1)-make)
+else
+$(1):image-dev-env go_build_cache_directory
+	@docker run $$$$(if [ -t 0 ] ;then echo -n "-it";fi) --rm --workdir /mnt/ \
+		--cap-add=sys_ptrace \
+		$(BUILD_INDOCKER_ARG) \
+		${DOCKER_REGISTRY_PREFIX}pingcap/dev-env:${IMAGE_TAG} \
+		/usr/bin/make $(1)
+endif
+endef
 
-# Generate code
-generate: $(GOBIN)/controller-gen chaos-build
-	cd ./api/v1alpha1 ;\
-		$< object:headerFile=../../hack/boilerplate/boilerplate.generatego.txt paths="./..." ;
+define proto-make
+	for dir in pkg/chaosdaemon pkg/chaoskernel ; do\
+		protoc -I $$$$dir/pb $$$$dir/pb/*.proto --go_out=plugins=grpc:$$$$dir/pb --go_out=./$$$$dir/pb ;\
+	done
+endef
+$(eval $(call RUN_IN_DEV_ENV_TEMPLATE,proto))
 
+define manifests/crd.yaml-make
+	kustomize build config/default > manifests/crd.yaml
+endef
+$(eval $(call RUN_IN_DEV_ENV_TEMPLATE,manifests/crd.yaml,config))
 
-manifests/crd.yaml: config ensure-kustomize
-	$(KUSTOMIZE_BIN) build config/default > manifests/crd.yaml
+define manifests/crd-v1beta1.yaml-make
+	mkdir ./output
```

Comment:
```suggestion
	mkdir -p ./output
```

Reply:
will be failed if `./output` already existed.

Reply:
nice catch !